extensions/amp-access/0.1/signin.js

```diff
@@ -0,0 +1,225 @@
+ * viewer and the source origin must all explicitly opt-in into this system.
+ * The source origin opts in by setting sign-in parameters in amp-access
+ * configuration. The viewer opts in by supplying `#signin=1` viewer parameter
+ * and user opt-in must be ensured by the viewer.
+ *
+ * There are two sub-protocols: access token and request sign-in.
+ *
+ * Access token is configured via `acceptAccessToken: true` configuration
+ * option. If the viewer can be asked for the access token in this case via
+ * `getAccessTokenPassive` message. Most likely, the viewer has exchanged
+ * the access token with the source origin using OAuth2 mechanism or similar.
+ * No special security measures are applied in AMP Runtime since the token is
+ * expected to be encrypted or signed by the source origin.
+ *
+ * Login dialog may return an access grant via `#access_grant` in the hash
+ * response. When `acceptAccessToken: true` is specified, the viewer may be
+ * asked to exchange this access grant for an access token using the
+ * `storeAccessToken` message.
+ *
+ * Request sign-in is configured via `signinServices: []` configuration option.
+ * E.g. `signinServices: ["https://accounts.google.com"]`. By using this option,
+ * the source origin states that it's ready to accept ID tokens produced by
+ * this authority. The viewer may be asked to implement sign-in with ID token
+ * via the `requestSignIn` message. The viewer must ensure the user's consent
+ * before sending any tokens to the source origin.
+ */
+export class SignInProtocol {
+
+  /**
+   * @param {!Window} win
+   * @param {!Viewer} viewer
+   * @param {string} pubOrigin
+   * @param {!JSONObject} configJson
+   */
+  constructor(win, viewer, pubOrigin, configJson) {
+
+    /** @const {!Window} */
+    this.win = win;
+
+    /** @private @const {!Viewer} */
+    this.viewer_ = viewer;
+
+    /** @private @const {string} */
+    this.pubOrigin_ = pubOrigin;
+
+    /** @private @const {boolean} */
+    this.isEnabled_ =
+        isExperimentOn(this.win, TAG) &&
+        this.viewer_.isEmbedded() &&
+        this.viewer_.getParam('signin') == '1';
+
+    if (this.isEnabled_) {
+
+      /** @private @const {boolean} */
+      this.acceptAccessToken_ = !!configJson['acceptAccessToken'];
+
+      const viewerSignInService = this.viewer_.getParam('signinService');
+      const configSignInServices = configJson['signinServices'];
+      if (configSignInServices) {
+        user.assert(isArray(configSignInServices),
+            '"signinServices" must be an array');
+      }
+
+      /** @private @const {boolean} */
+      this.supportsSignInService_ = configSignInServices &&
+          configSignInServices.indexOf(viewerSignInService) != -1;
+
+      /** @private {!Promise<?string>} */
+      this.accessTokenPromise_ = null;
+    }
+  }
+
+  /**
+   * Whether this protocol has been enabled by the viewer.
+   * @return {boolean}
+   */
+  isEnabled() {
+    return this.isEnabled_;
+  }
+
+  /**
+   */
```

Comment:
I added docs. But generally, any method should have jsdoc, I believe. Even if empty.